src/VirtualContent.js

```diff
@@ -0,0 +1,341 @@
+          this[_resizeObserver].unobserve(node);
+          this[_intersectionObserver].unobserve(node);
+          node.removeAttribute('invisible');
+          estimatedHeights.delete(node);
+        }
+      }
+
+      for (const node of record.addedNodes) {
+        if (node.nodeType === Node.ELEMENT_NODE) {
+          // Added children should be invisible initially.
+          node.setAttribute('invisible', '');
+          estimatedHeights.set(node, DEFAULT_HEIGHT_ESTIMATE);
+        } else {
+          // Remove non-element children because we can't control their
+          // invisibility state or even prevent them from being rendered using
+          // CSS (they aren't distinctly selectable).
+
+          // These records are not coalesced, so test that the node is actually
+          // a child of this node before removing it.
+          if (node.parentNode === this) {
+            this.removeChild(node);
+          }
+        }
+      }
+    }
+
+    this[_scheduleUpdate]();
+  }
+
+  [_resizeObserverCallback]() {
+    this[_scheduleUpdate]();
+  }
+
+  [_onActivateinvisible](e) {
+    // Find the child containing the target and synchronously update, forcing
+    // that child to be visible. The browser will automatically scroll to that
+    // element because it is visible, which will trigger another update to make
+    // the surrounding nodes visible.
+    const child = e.target;
+    while (child.parentNode !== this) {
+      child = child.parentNode;
+    }
+    this[_update]({forceVisible: new Set([child])});
+  }
+
+  [_scheduleUpdate]() {
+    if (this[_updateRAFToken] !== undefined) return;
+
+    this[_updateRAFToken] = window.requestAnimationFrame(this[_update]);
+  }
+
+  [_update]({forceVisible = new Set()} = {}) {
+    this[_updateRAFToken] = undefined;
+
+    const thisClientRect = this.getBoundingClientRect();
+    // Don't read or store layout information if this element isn't in a
+    // renderable state (e.g. disconnected, invisible, `display: none`, etc.).
+    const isRenderable = !(
+      thisClientRect.top === 0 &&
+      thisClientRect.left === 0 &&
+      thisClientRect.width === 0 &&
+      thisClientRect.height === 0
+    );
+
+    const estimatedHeights = this[_estimatedHeights];
+    const updateHeightEstimate = (child) => {
+      if (isRenderable && !child.hasAttribute('invisible')) {
+        const childClientRect = child.getBoundingClientRect();
+        const style = window.getComputedStyle(child);
+        const height =
+          window.parseFloat(style.marginTop, 10) +
+          window.parseFloat(style.marginBottom, 10) +
+          childClientRect.height;
+        estimatedHeights.set(child, height);
+      }
+      return estimatedHeights.get(child);
+    };
+
+    const previouslyVisible = new Set();
+    for (let child = this.firstChild; child !== null; child = child.nextSibling) {
+      if (!child.hasAttribute('invisible')) {
+        previouslyVisible.add(child);
+      }
+    }
```

Comment:
I think a simpler version would be

```js
const previouslyVisible = new Set(this.querySelector('> *:not([invisible])'));
```

But maybe this is faster.

Reply:
Unfortunately, relative selectors aren't supported by `querySelector{,All}` afaik, but I don't there's any good reason why I didn't at least try for something similar. I think I just didn't consider it since I was already iterating below. Maybe `Array.from(this.querySelectorAll('*:not([invisible])')).filter(node => node.parentNode === this)` would be worth trying out? Or adding some random garbage attribute for the sake of adding it to the selector so that the filtering bit can be dropped?